### PR TITLE
PUD-1598  Implement API for part B records

### PR DESCRIPTION
**Pull request by AndyNewton**
Should allow Jon's UI branch to be merged.  Multiple error handling for virus found to follow as PUD-1605.  Minor feedback updates I'll include with that.  OK @philsegal ?

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/PartBRecordController.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  ): PartBRecordId =
+    tokenExtractor.getTokenFromHeader(bearerToken).userUuid().let { currentUserId ->
+      partBRecordService.createRecord(recallId, currentUserId, request).map {
+        it
+      }.onFailure {
+        throw VirusFoundException()
+      }
+    }
+
+  data class PartBRequest(
```
Comment (philsegal) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/PartBRecordController.kt:47]: Maybe a PartBRecordRequest, but 🤷‍♂️

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallController.kt [hunk 2]
```diff
@@ -408,6 +426,7 @@ data class RecallResponse(
   val documents: List<Api.RecallDocument> = emptyList(),
   val lastKnownAddresses: List<Api.LastKnownAddress> = emptyList(),
   val missingDocumentsRecords: List<Api.MissingDocumentsRecord> = emptyList(),
+  val partBRecords: List<Api.PartBRecord> = emptyList(),
```
Comment (philsegal) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallController.kt:429]: alphabetical please 🙏

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/db/PartBRecord.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+  @Column(name = "part_b_document_id", nullable = false)
+  @JoinColumn(name = "document_id", table = "document")
+  val partBDocumentId: UUID,
+
+  @Column(nullable = false)
+  @JoinColumn(name = "document_id", table = "document")
+  val emailId: UUID,
+
+  @Column(nullable = true)
```
Comment (philsegal) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/db/PartBRecord.kt:39]: not sure you need `nullable = true`, but 🤷‍♂️

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/domain/domain.kt [hunk 4]
```diff
@@ -41,6 +41,7 @@ class PrisonId(value: String) : Validated<String>(value, notBlank, alphanumeric,
 class PrisonName(value: String) : Validated<String>(value, notBlank)
 class RecallId(value: UUID) : Validated<UUID>(value)
 class RescindRecordId(value: UUID) : Validated<UUID>(value)
+class PartBRecordId(value: UUID) : Validated<UUID>(value)
```
Comment (philsegal) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/domain/domain.kt:44]: alphabetical 😄

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    request: PartBRequest
+  ): Result<PartBRecordId, List<Pair<DocumentCategory, FileName>>> =
+    recallRepository.getByRecallId(recallId).let { recall ->
+      // TODO: Move validation to DTO?  Even then, cross-field validation requires custom code.
+      //  As these properties are mutually dependent an alternative would be to combine them as a single object in the DTO
+      val oasysName = request.oasysFileName
+      val oasysContent = request.oasysFileContent
+      if ((oasysName == null && !oasysContent.isNullOrBlank()) || (oasysContent.isNullOrBlank() && oasysName != null)) {
+        throw IllegalArgumentException("Both fileName and fileContent of OASYS must be supplied as non-blank or neither")
+      }
```
Comment (philsegal) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt:49]: should/could this be on the record itself?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  private fun scanAndStore(
+    category: DocumentCategory,
+    details: String,
+    fileContent: String,
+    fileName: FileName,
+    recallId: RecallId,
+    currentUserId: UserId
+  ): Pair<DocumentCategory, Pair<FileName, Result<DocumentId, VirusScanResult>>> =
```
Comment (philsegal) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt:100]: 🤮 . Does it need the filename too? The Pair -> Pair feels nasty.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/PartBRecordComponentTest.kt [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    assertThat(recallWithPartBRecord.partBRecords.first().partBRecordId, equalTo(partBRecordId))
+    assertThat(recallWithPartBRecord.partBRecords.first().details, equalTo(details))
+    assertThat(recallWithPartBRecord.partBRecords.first().partBReceivedDate, equalTo(emailReceivedDate))
+    assertThat(recallWithPartBRecord.partBRecords.first().partBDocumentId, present())
+    assertThat(recallWithPartBRecord.partBRecords.first().partBFileName, equalTo(partBFileName))
+    assertThat(recallWithPartBRecord.partBRecords.first().emailId, present())
+    assertThat(recallWithPartBRecord.partBRecords.first().emailFileName, equalTo(emailFileName))
+    assertThat(recallWithPartBRecord.partBRecords.first().oasysDocumentId, present())
+    assertThat(recallWithPartBRecord.partBRecords.first().oasysFileName, equalTo(oasysFileName))
+    assertThat(recallWithPartBRecord.partBRecords.first().version, equalTo(1))
```
Comment (philsegal) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/PartBRecordComponentTest.kt:67]: fixedClock might make this more concise?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/pact/provider/ManageRecallsUIPactTest.kt [hunk 8]
```diff
@@ -236,6 +244,16 @@ class ManagerRecallsUiAuthorizedPactTest : ManagerRecallsUiPactTestBase() {
     `a note exists`(2)
   }
 
+  @State(
+    "a recall with part B records exists",
+    "a Fixed Term recall with part B records exists",
```
Comment (philsegal) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/pact/provider/ManageRecallsUIPactTest.kt:249]: not sure this makes sense. Isn't Part B only for Standard recalls?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordServiceTest.kt [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+    val response = underTest.createRecord(recallId, userId, request)
+
+    assertThat(response, equalTo(Success(partBRecordId)))
+    assertThat(savedPartBRecordSlot.captured.details, equalTo(details))
+    assertThat(savedPartBRecordSlot.captured.partBReceivedDate, equalTo(emailReceivedDate))
+    assertThat(savedPartBRecordSlot.captured.partBDocumentId, equalTo(partBDocumentId.value))
+    assertThat(savedPartBRecordSlot.captured.emailId, equalTo(emailDocumentId.value))
+    assertThat(savedPartBRecordSlot.captured.oasysDocumentId, equalTo(oasysDocumentId.value))
+    assertThat(savedPartBRecordSlot.captured.version, equalTo(1))
+    assertThat(savedPartBRecordSlot.captured.createdByUserId, equalTo(userId.value))
```
Comment (philsegal) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordServiceTest.kt:86]: fixedClock?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt (hunk 6 above)
Comment (AndyNewton) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt:100]: Jon wants the filename and category returned in the event of error: easiest way to do that looks like keeping them together in the returned Failure 🤷‍♂️ .

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/PartBRecordController.kt (hunk 1 above)
Comment (AndyNewton) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/PartBRecordController.kt:47]: Will consider.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallController.kt (hunk 2 above)
Comment (AndyNewton) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallController.kt:429]: Will do.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/db/PartBRecord.kt (hunk 3 above)
Comment (AndyNewton) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/db/PartBRecord.kt:39]: Probably not; more explicit though?  🤷‍♂️

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/domain/domain.kt (hunk 4 above)
Comment (AndyNewton) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/domain/domain.kt:44]: Will do.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt (hunk 5 above)
Comment (AndyNewton) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt:49]: I have a // TODO above it to consider how else to apply validation but maybe swings and roundabouts.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/PartBRecordComponentTest.kt (hunk 7 above)
Comment (AndyNewton) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/PartBRecordComponentTest.kt:67]: 🤔

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/pact/provider/ManageRecallsUIPactTest.kt (hunk 8 above)
Comment (AndyNewton) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/pact/provider/ManageRecallsUIPactTest.kt:249]: Will bin.  TBH I'm not convinced we need any of these "a Fixed Term recall blah ..." states?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordServiceTest.kt (hunk 9 above)
Comment (AndyNewton) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordServiceTest.kt:86]: 🤔

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt (hunk 6 above)
Comment (philsegal) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt:100]: Ok, not sure why you'd need the filename

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/pact/provider/ManageRecallsUIPactTest.kt (hunk 8 above)
Comment (philsegal) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/pact/provider/ManageRecallsUIPactTest.kt:249]: I think only reason to have them is if there is something specific about the state which means it must be Fixed/Standard. but agree, possibly gone overkill

**Review by philsegal — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/db/PartBRecord.kt (hunk 3 above)
Comment (AndyNewton) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/db/PartBRecord.kt:39]: Have removed all across all entity classes.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt (hunk 5 above)
Comment (AndyNewton) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt:49]: Only the document Id is on the record so there's no scope there for validating name + content together.  I've added this to the // TODO: [PUD-1626](https://dsdmoj.atlassian.net/browse/PUD-1626) Cross-property Request validation: move to DTOs or restructure Request classes?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt (hunk 6 above)
Comment (AndyNewton) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt:100]: https://mojdt.slack.com/archives/C0223AGGQU8/p1648026203846159?thread_ts=1647360199.283689&cid=C0223AGGQU8

**Pull request merged**